### PR TITLE
Update: Move methods to SourceCode (fixes #3516)

### LINKED ISSUE #3516: Move some eslint methods to SourceCode
Now that `SourceCode` is merged, I'd like to move more of the AST-specific functionality to it. The `eslint` object right now has a mix of methods such as `getComments()` that apply only to the AST and really should never have been on the object.

My goal is to move these methods to `SourceCode` and then have the methods on `eslint` just delegate to `SourceCode`. We can't really remove those methods since others might be relying on them, but we can certainly refactor so that the code doesn't live inside of the `eslint` object.

### PR DESCRIPTION
**Pull request by nzakas**
Main goals of this direction:

1. Remove code from `lib/eslint.js`
1. Make `SourceCode` the owner of all things related to nodes and tokens
1. Prevent method bloat on `context` by moving most methods to `SourceCode`

### CODE REVIEW

**Comment by nzakas** [comment]
Now ready for review.

**Comment by gyandeeps** [comment]
LGTM. 
Do we want to move the unit test of all methods which moved from `eslint` to `SourceCode` module to its new test module?

**Comment by nzakas** [comment]
Yeah, that's probably a good idea.

File: docs/developer-guide/working-with-rules.md [hunk 1]
```diff
@@ -76,30 +76,35 @@ The `context` object contains additional functionality that is helpful for rules
...
-* `getTokens(node)` - returns all tokens for the given node.
-* `getTokensAfter(nodeOrToken, count)` - returns `count` tokens after the given node or token.
-* `getTokensBefore(nodeOrToken, count)` - returns `count` tokens before the given node or token.
-* `getTokensBetween(node1, node2)` - returns the tokens between two nodes.
+* `getSourceCode()` - returns a `SourceCode` object that you can use to work with the source that was passed to ESLint
+* `isMarkedAsUsed(name)` - returns true if a given variable name has been marked as used.
 * `markVariableAsUsed(name)` - marks the named variable in scope as used. This affects the [no-unused-vars](../rules/no-unused-vars.md) rule.
 * `report(node, message)` - reports an error in the code.
 
+**Deprecated:** The following methods on the `context` object are deprecated. Please use the corresponding methods on `SourceCode` instead:
```
Comment (BYK) [inline docs/developer-guide/working-with-rules.md]: Should we call this simply `Source`?

File: lib/eslint.js [hunk 2]
```diff
@@ -940,28 +890,33 @@ module.exports = (function() {
...
-     * Retrieves the JSDoc comment for a given node.
-     * @param {ASTNode} node The AST node to get the comment for.
-     * @returns {ASTNode} The BlockComment node containing the JSDoc for the
-     *      given node or null if not found.
-     */
-    api.getJSDocComment = astUtils.getJSDocComment;
+    // copy over methods
+    Object.keys(externalMethods).forEach(function(methodName) {
+        api[methodName] = function() {
+            return sourceCode ? sourceCode[externalMethods[methodName]].apply(sourceCode, arguments) : null;
```
Comment (BYK) [inline lib/eslint.js]: Why not use `bind` here?

**Comment by BYK** [comment]
LGTM.

We gotta plan/ticket moving all other rules to this new stuff.

File: docs/developer-guide/working-with-rules.md (hunk 1 above)
Comment (nzakas) [inline docs/developer-guide/working-with-rules.md]: We already have `context.getSource()`, so calling this `Source` would be confusing.

File: lib/eslint.js (hunk 2 above)
Comment (nzakas) [inline lib/eslint.js]: There's an odd case where `sourceCode` can be null, and in that case, we cannot throw an error. So `bind()` wouldn't work for all cases here.

Comment (BYK) [inline lib/eslint.js]: Oh right. Reading the code carfully, I see how this makes a difference now.

**Comment by ilyavolodin** [comment]
LGTM. @BYK It looks like @nzakas already did most if not all of the rules.

**Comment by nzakas** [comment]
I only did for the ast-utils piece. If like to start converting other rules so they are not using all the methods on `context`.

**Pull request merged**